msm/msm

```diff
@@ -439,61 +439,120 @@ function search() {
+  #       mycroft-core.
+  if [ ! -f /tmp/mycroft_has_re-PIPed ] ; then
+     run_pip "$1"
+     echo "Re-running PIP on requirements.txt"
+  fi
+}
+
+function update_skill() {
+  local folder=$1
+
+  cd "$folder"
+  ignore_pyc
+  if ! should_update_git; then
+    echo "Skipped $d."
+    return
+  fi
+
+  requirements_hash() {
+    [ ! -f requirements.txt ] || git rev-parse HEAD:requirements.txt
+    [ ! -f requirements.sh ] || git rev-parse HEAD:requirements.sh
+  }
+
+  before_head=$(git rev-parse HEAD)
+  before=$(requirements_hash)
+
+  output=$(git fetch && git reset --hard origin/master)
+
+  after=$(requirements_hash)
+  after_head=$(git rev-parse HEAD)
+
+  if [ "$before_head" != "$after_head" ]; then
+    echo "$output"
+    echo "Updated $folder."
+  else
+    echo "Checked $folder."
+  fi
+
+  if [ "$before" != "$after" ]; then
+    echo "Updating dependencies..."
+    run_pip "$folder"
+  fi
+
+  pip_hack "$folder"
+}
+
 function update() {
+  workon mycroft &>/dev/null || true
+  
   echo "Updating installed skills..."
   cd "${mycroft_skill_folder}"
+  
+  send_start_update & 
+  pids="$!"
 
   # Loop through all of the current Skill folders
   for d in $(find "${mycroft_skill_folder}" -mindepth 1 -maxdepth 1 -type d |grep -v '.git'$ ); do
     # Go in to all folders that are git checkouts
-    if git -C "$d" rev-parse --git-dir > /dev/null 2>&1; then
-      cd "${d}"
-      UPSTREAM=${1:-'@{u}'}
-      LOCAL=$(git rev-parse @)
-      REMOTE=$(git rev-parse "$UPSTREAM")
-      BASE=$(git merge-base @ "$UPSTREAM")
-
-      # Force ignoring the generated .pyc files
-      if ! grep -q '.pyc'$ .git/info/exclude; then
-        echo "*.pyc" >> .git/info/exclude
-      fi
-
-      BRANCH="$(git symbolic-ref HEAD 2>/dev/null)"
-      BRANCH="${BRANCH##refs/heads/}"
-
-      # Only update checkouts that have not been modified at all
-      if [[ (-z $(git status --porcelain --untracked-files=no)) &&  # No Modified files
-            !($LOCAL != $REMOTE && $REMOTE = $BASE) &&  # No new commits
-            "$BRANCH" = "master" ]]  # On master branch
-      then
-        echo "Updating ${d}..."
-        echo -n "  "
-        send_start_update
-        git fetch
-        git reset --hard origin/master
-        rm -f *.pyc
-        run_pip "${d}"
-      else
-        echo "Ignoring ${d}, skill has been modified."
-      fi
-
-      # TODO: Remove this for 18.02
-      # HACK: Re-run PIP, because Mycroft-core removed some of its required,
-      #       which broke previously-installed packages on a Picroft/Mark1
-      #       that was satisfied by packages that previously came along with
-      #       mycroft-core.
-      if [ ! -f /tmp/mycroft_has_re-PIPed ] ; then
-         run_pip "${d}"
-         echo "Re-running PIP on requirements.txt"
-      fi
+    git_dir=$(git -C "$d" rev-parse --git-dir 2>&1)
+    if [ "$git_dir" = ".git" ]; then
+      update_skill "$d" &
+      pids="$pids $!"
     fi
   done
 
+  wait $pids
```

Comment:
The wait statement doesn't seem to work quite right:
```
./msm/msm: line 548: wait: `22635 22642 22647 22657 22667 22676 22683 22694 22707 22714 22721 22732 22746 22754 22762 22770 22778 22793 22804 22815 22828 22837 22856 22869 22879 22910 22932 22945 22957 22974 22986 23001 23011 23028 23038 23054 23066 23077 23096 23124 23133': not a pid or valid job spec
```

And no wait occurs.

Reply:
Hmm, seems to be a difference between out bash versions. See if the change I made fixes this for you.